### PR TITLE
rendering optimization

### PR DESCRIPTION
**Pull request by horpto**
I gain a huge speedup, rendering of table in terminal mode decreased from 30s to 0.5s.
I'm sure there are still places for optimization, for example rendering of cell.

### CODE REVIEW

**Comment by horpto** [comment]
@krobertson 
done

**Comment by horpto** [comment]
@krobertson 
ping?

**Comment by johnSchnake** [comment]
Hey. I great change, sorry for the big delay.  There's a conflict now but after resolving I dont see any reason not to include this. The `cell.formattedValue` isn't being updated anywhere so we should definitely cache that width.  Maybe just add a comment on the object definition that the width is being cached and should be updated manually if the internal contents are changed.

File: style.go [hunk 1]
```diff
@@ -198,6 +179,24 @@ func createRenderStyle(table *Table) *renderStyle {
 	return style
 }
 
+func (s *renderStyle) calculateCellWidths(elements []Element) {
+	// loop over the rows and cells to calculate widthss
```
Comment (johnSchnake) [inline style.go]: Typo; `widths`

File: table.go [hunk 2]
```diff
@@ -217,61 +218,72 @@ func (t *Table) Render() (buffer string) {
...
-		if t.title != nil && t.headers == nil {
-			t.elements = append([]Element{&Separator{where: LINE_SUBTOP}}, t.elements...)
-		} else if t.title == nil && t.headers == nil {
-			t.elements = append([]Element{&Separator{where: LINE_TOP}}, t.elements...)
-		} else {
-			t.elements = append([]Element{&Separator{where: LINE_INNER}}, t.elements...)
-		}
+func (t *Table) renderTerminal() string {
+	// save elements that should be before rows
+
```
Comment (johnSchnake) [inline table.go]: nit: remove empty line.

**Comment by johnSchnake** [comment]
We got huge speedups when fixing some of the code to use byte buffers. If you want to try to rebase this and see a big speedup in the benchmarks let us know. Closing otherwise.

**Pull request closed without merging**